diesel_codegen_shared/src/schema_inference/mod.rs

```diff
@@ -0,0 +1,54 @@
+#[cfg(feature = "postgres")]
+mod pg;
+#[cfg(feature = "sqlite")]
+mod sqlite;
+
+use diesel::Connection;
+use std::error::Error;
+
+use database_url::extract_database_url;
+use InferConnection;
+
+pub fn load_table_names(database_url: &str) -> Result<Vec<String>, Box<Error>> {
+    let connection = try!(establish_connection(database_url));
+
+    match connection {
+        #[cfg(feature = "sqlite")]
+        InferConnection::Sqlite(c) => sqlite::load_table_names(&c),
+        #[cfg(feature = "postgres")]
+        InferConnection::Pg(c) => pg::load_table_names(&c),
+    }
+}
+
+fn establish_connection(database_url: &str) -> Result<InferConnection, Box<Error>> {
+    let database_url = try!(extract_database_url(database_url));
+    match database_url {
+        #[cfg(feature = "postgres")]
+        _ if database_url.starts_with("postgres://") || database_url.starts_with("postgresql://") => {
```

Comment:
imho, `match` with `_ if` is a code smell, but maybe it's easier to do the `cfg`s this way?